src/OAuth2.php

```diff
@@ -703,6 +703,26 @@ public function getCacheKey()
         return null;
     }
 
+    /**
+     * Obtains the scope or the Audience and formats it for external classes to use
+     * in generating a cache key.
+     *
+     * @return ?string a key that may be used to cache the auth token.
+     */
+    public function getFormattedScopeOrAudience()
```

Comment:
Since this is a utility class, I'd prefer to not expose a public method like this on it (which is really an internal helper method). 

What if we added this method to `CacheTrait`, made it private, and made it accept `$scope` and `$audience` parameters.  Then we'd call it like

```php
return $this->auth->getClientId() . ':' . $this->getFormattedScopeOrAudience($this->auth->getScope(), $this->auth->getAudience());
```

Or even skip this method all together and just have the cache keys look like this (simpler):

```php
return $this->auth->getClientId() . ':' . implode(':', $this->auth->getScope()) . ':' . $this->auth->getAudience();
```

Reply:
Yeah I was also on the fence with this one, don't think formatting inside `OAuth` is a responsability of that class.

Good idea, let me change it.

Reply:
On a second thought, while reviewing this, none of the credential classes uses the `CacheTrait` and it has some methods that I don't believe belong on these classes.

For example: `getCachedValue` and `setCacheValue`. These methods are currently being used by the `fetchAuthTokenCache`. I mention that these methods shouldn't be on the credentials themselves but I should stress they shouldn't be on this specific implementation as in V2 would be a different structure. But I digress.

I am not a fan of writing explicit code over and over, it always tingles my "This should be abstracted" sense, perhaps another trait? A utility static class?. That being said I also don't think is a do or die situation in this case.

Reply:
I was worried about that (wrt the `CacheTrait`). In V2 all these classes will implement `CacheTrait`, and we'll get rid of `FetchAuthTokenCache` (although we _could_ that that now, without breaking BC, and just have `FetchAuthTokenCache` be a noop.... but I digress!)

At the end of the day it's a matter of preference. You could put it in `CredentialsLoader`, as all the credentials  in question extend that?  

Generally speaking, for simple things such as concatenating strings as we are doing here, which aren't SUPER important to keep consistent, I lean towards code repetition. To me this is clearer and easier to read than an abstraction / redirection in the code. But if we can cleanly add dryness as well (which I think we can with `CredentialsLoader`), then I'm okay with it.

Reply:
I agree with your statement of readability. `ExternalAccountCredentials` do not implement `CredentialsLoader` but we just had a discussion about how different that one is that all the `sources` classes are doing their own implementation so we could add it to the `CredentialsLoader`.

Still, will go with the code repetition.